toolchain/check/context.cpp

```diff
@@ -216,6 +216,14 @@ auto Context::DiagnoseDuplicateName(SemIRLoc dup_def, SemIRLoc prev_def)
       .Emit();
 }
 
+auto Context::DiagnosePoisonedName(SemIRLoc dup_def) -> void {
+  CARBON_DIAGNOSTIC(
+      NameDeclPoisoned, Error,
+      "cannot declare this name in this scope since is was already used "
+      "without qualification in the context of this scope");
```

Comment:
We have a little [diagnostic style guide](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/diagnostics.md#diagnostic-message-style-guide). Generally I think we're trying to focus on the situation the toolchain observed first. Suggesting a rephrasing which I _think_ better reflects the intent.

Reply:
I wonder if it'd be better to put the diagnostic at the point of use, and say something like "name used before it was declared" with a note pointing to the later declaration?

Reply:
That seems reasonable, though noting [my comments about location above](https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1876530039), so for now this would probably look like using an `Invalid` ID for the location that needs to be piped through:

```
CARBON_DIAGNOSTIC(NameUseBeforeDecl, Error, "name used before it was declared");
CARBON_DIAGNOSTIC(NameUseBeforeDeclNote, Note, "declared here");
// TODO: This should pass in the original declaration.
unit_info.emitter.Emit(Parse::NodeId::Invalid, NameUseBeforeDecl)
    .Note(decl_loc, NameUseBeforeDeclNote)
    .Emit();
```

Reply:
Thanks!
I've changed the diagnostic name and text.
It makes sense to put the warning where the name is poisoned, though I find an error in code that is caused because of code that comes later somewhat confusing.
I'm ok with both options.
Anyways, this needs to wait for when we store the data on what caused the poisoning.

Reply:
Unresolving -- I think we should probably go with the phrasing zygoloid suggested, with TODOs per https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1876772138

Reply:
Right now we have the location of the later declaration.
IIUC, the suggestion is to error where the poisoning happened with a note on the later declaration.
If we go with this suggestion, we only have the note, not the error.
Can you clarify what you suggest for this PR?
I've made the TODO clearer.

Reply:
Sure, I've copied in my code from the above comment into a suggested edit: https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1880640674.

Reply:
Thanks for the clarification.
Done.
See my reply in the other comment.